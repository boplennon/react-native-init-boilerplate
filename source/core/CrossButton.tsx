import React from 'react';
// @ts-ignore
import { Button } from 'react-native-paper';
import { Icon } from 'react-native-elements';
import Theme from '../Styles';
import { ViewStyle, StyleProp, View, ButtonProps } from 'react-native';
import { Color } from 'csstype';
import _ from 'lodash';

/**
 * Extends button props. Remarks: use custom {@link onPressButton} event
 */
export interface IButtonProps extends ButtonProps {
  textColor?: Color;
  iconName?: string;
  backgroundColor?: Color;
  style?: StyleProp<ViewStyle>;
  /**
   * Custom pressed event to use instead of `onPress`
   */
  onPressButton?: () => void;
}

/**
 * A custom button that displays as an icon if {@param title} is not supplied
 *
 * Remarks: use custom {@param onPressButton} event
 * @param param0 params
 */
export class CrossButton extends React.Component<IButtonProps> {
  render() {
  return (
    <View style={Theme.container}>
      {_.isNil(this.props.title) ? (
        <Icon
          reverse
          {...this.props}
          onPress={this.props.onPressButton ? this.props.onPressButton : () => console.log('CrossButton')}
          name={_.isNil(this.props.iconName) ? 'house' : this.props.iconName.toString()}
          type='font-awesome'
          color={this.props.backgroundColor}
        />
      ) : (
        <Button
          color={this.props.backgroundColor}
          {...this.props}
          style={this.props.style || Theme.button}
          onPress={this.props.onPressButton ? this.props.onPressButton : () =>  console.log('CrossButton')}
          mode='contained'
        >
          {this.props.title}
        </Button>
      )}
    </View>
  ); }
}

export default CrossButton;
