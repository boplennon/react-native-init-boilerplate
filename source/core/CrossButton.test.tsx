// core/components/BusyIndicator.test
/// <reference types="jest"/>
import React from 'react';
import { shallow } from 'enzyme';
import CrossButton from './CrossButton';

jest.unmock('react-native');

function setup(iconName:string, title:string) {

  const enzymeWrapper = shallow(<CrossButton iconName={iconName} title={title} onPress={jest.fn()}/>);

  return {
    enzymeWrapper,
  };
}

describe('components', () => {
  /**
   * Test component rendering. Properties of children might be tested by importing their type:
   *
   *    const busyProps = enzymeWrapper.find(BusyIndicator).props();
   *    expect(busyProps.isBusy).toBe(false);
   */
  describe('CrossButton', () => {
    it('With title should render', () => {
      const { enzymeWrapper } = setup('home', 'home');
      expect(enzymeWrapper).toMatchSnapshot();
    });

    it('Without title should renders', () => {
        const { enzymeWrapper } = setup('map', null);
        expect(enzymeWrapper).toMatchSnapshot();
      });
  });
});
