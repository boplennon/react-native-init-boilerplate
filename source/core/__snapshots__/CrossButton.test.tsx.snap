// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CrossButton With title should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CrossButton
    iconName="home"
    onPress={[MockFunction]}
    title="home"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <UNDEFINED
        iconName="home"
        mode="contained"
        onPress={[Function]}
        style={
          Object {
            "height": 45,
            "marginBottom": 10,
            "marginRight": 10,
            "marginTop": 10,
          }
        }
        title="home"
      >
        home
      </UNDEFINED>,
      "style": Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "home",
        "color": undefined,
        "iconName": "home",
        "mode": "contained",
        "onPress": [Function],
        "style": Object {
          "height": 45,
          "marginBottom": 10,
          "marginRight": 10,
          "marginTop": 10,
        },
        "title": "home",
      },
      "ref": null,
      "rendered": "home",
      "type": undefined,
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <UNDEFINED
          iconName="home"
          mode="contained"
          onPress={[Function]}
          style={
            Object {
              "height": 45,
              "marginBottom": 10,
              "marginRight": 10,
              "marginTop": 10,
            }
          }
          title="home"
        >
          home
        </UNDEFINED>,
        "style": Object {
          "flex": 1,
          "justifyContent": "space-between",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "home",
          "color": undefined,
          "iconName": "home",
          "mode": "contained",
          "onPress": [Function],
          "style": Object {
            "height": 45,
            "marginBottom": 10,
            "marginRight": 10,
            "marginTop": 10,
          },
          "title": "home",
        },
        "ref": null,
        "rendered": "home",
        "type": undefined,
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`components CrossButton Without title should renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CrossButton
    iconName="map"
    onPress={[MockFunction]}
    title={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <View
        iconName="map"
        name="map"
        onPress={[Function]}
        reverse={true}
        title={null}
        type="font-awesome"
      />,
      "style": Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "color": undefined,
        "iconName": "map",
        "name": "map",
        "onPress": [Function],
        "reverse": true,
        "title": null,
        "type": "font-awesome",
      },
      "ref": null,
      "rendered": null,
      "type": "View",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <View
          iconName="map"
          name="map"
          onPress={[Function]}
          reverse={true}
          title={null}
          type="font-awesome"
        />,
        "style": Object {
          "flex": 1,
          "justifyContent": "space-between",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "color": undefined,
          "iconName": "map",
          "name": "map",
          "onPress": [Function],
          "reverse": true,
          "title": null,
          "type": "font-awesome",
        },
        "ref": null,
        "rendered": null,
        "type": "View",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
